tooltip_scheme <- "Describes the set of codons used to encode the peptides.<br><br>PeLiCa considers several different encoding schemes. Peptide encoding sequences that contain stop codons will, in general, not lead to functional library particles. They are therefore treated as invalid sequences by PeLiCa. Cysteines incorporated in random peptides, were described to inhibit the generation of peptide bearing phages. Further, a strong underrepresentation of cysteine containing peptides observed in peptide libraries displayed on the surface of different AAV vectors suggests that cysteins can also have a negative effect in at least some viral systems. Notably, there is also a distinct lack of naturally occurring capsid surface-exposed cysteine residues on wild type AAV2 and human Adenovirus type 5. Therefore, cysteine containing peptides might be regarded as invalid sequences in relevant display systems. For a detailed explanation and relevant references, please refer to our publication: Sieber T, Hare E, Hofmann H, and Trepel M; Biomathematical description of peptide library properties, Nucleic Acids Research 2014.<br><br>To reflect this, PeLiCa offers pre-set encoding schemes that allow you to treat cysteine-containing peptides as valid (NNN, NNB, NNK/S, 20/20), or invalid (NNN (-C), NNB (-C), NNK/S (-C) and 20/20 (-C)).<br><br>-- NNN: All four bases (\"N\" = G/A/T/C) possible at all three positions in the codon.<br>-- NNB: All four bases in the first two codon positions possible, the third position is restricted to G, T or C (= \"B\")<br>-- NNK/S: All four bases in the first two codon positions possible, the third position is restricted to G/T (= \"K\") or two C/G (= \"S\"). Though both schemes differ biologically, they behave equivalent in the analyses presented here.<br>-- 20/20: 20/20 describes the concept that DNA is assembled from prefabricated trimeric building blocks. This allows the generation of libraries from a predefined set of codons and thereby complete exclusion of Stop codons and other unwanted codons.<br>-- NNN (-C): NNN with Cysteine ignored.<br>-- NNB (-C): NNB with Cysteine ignored.<br>-- NNK/SC (-C): NNK/S with Cysteine ignored.<br>-- 20/20 (-C): 20/20 with Cysteine ignored.<br>-- Custom: Upload your own custom encoding scheme.<br><br>The schemes differ in the number of used codons, ranging from 64 (NNN), 48 (NNB), 32 (NNK/S) to 20 or less (20/20). Coding schemes that allow varying ratios of codons/amino acid, result in libraries biased towards amino acids which are encoded more often. Further, the number of Stop codons that can lead to premature termination of the peptide sequence influences the performance of the library."
tooltip_length <- "Input the number of amino acids in the library peptides. This can range from a minimum of 1 to a maximum of 20."
tooltip_size <- "Input the library size.  For instance, if your library contains 2.3 * 10^8 DNA clones, you would enter 2.3 in the first box, and 8 in the second box. This can range from a minimum of 1 * 10^1 to a maximum of 9.9 * 10^25."
tooltip_peptide <- "Here you can enter a specific peptide sequence (single letter code) and learn more about its performance within the library."
tooltip_custom <- "Instructions for using PeLiCa with a custom library scheme:<br><br>1. Create an Amino Acid class for each amino acid corresponding to the number of codons encoding each. For instance, if all amino acids have a single codon representation, you would put ACDEFHIKLMNPQRSTVWY in the Amino Acids box, and 1 in the Codon box.<br>2. Add a new AA class for each unique number of codon representations in your library scheme.<br>3. The * character (asterisk) represents stop tags - Select how many codons encode stop tags in your library.<br>4. Press the Done button when complete, then wait while PeLiCa computes the properties of your custom library. Typically, this will take about a minute.<br>5. Upon successful completion, PeLiCa will incorporate your custom scheme into the results and plots."
tooltip_digits <- "Enter the number of significant digits to use for values throughout PeLiCa"
tooltip_schemes <- "Check the box corresponding to library schemes you would like to display in the plots and tables"
tooltip_lengths <- "Check the box corresponding to peptide lengths you would like to display in the plots and tables"

tooltip_welcome <- "PeLiCA is best viewed using:<br><br>--Browser: Google Chrome, Mozilla Firefox, Safari, or Internet Explorer 10 (IE7, 8, and 9 are unsupported)<br>--Resolution: 1280x800 or higher"

tooltip_summary_library <- "Information on your selected library. Your selected configuration, along with values for the expected coverage, relative efficiency, peptide diversity, and functional diversity are displayed."
tooltip_summary_peptide <- "Information on your selected peptide, and the best and worst case peptides in your library. Neighborhood degrees are defined in terms of the number of conservative exchanges. A conservative exchange was defined here as any exchange that results in a positive log-odds score in the BLOSOM80 matrix (Henikoff, S. and Henikoff, J. G. (1992). Amino acid substitution matrices from protein blocks. Proceedings of the National Academy of Sciences, 89(22), 10915–10919.)"
tooltip_summary_scheme <- "Information on your selected scheme. The classes are defined in terms of the number of codon representations (c) for each amino acid (aacid). The s column is the total number of amino acids in each class."
tooltip_summary_sample <- "Sample of possible peptides from your library. The columns are as follows:<br><br>Peptide: Exemplary peptide sequence.<br>Peptide Class: Encoding class of the peptide (see: Biomathematical Description of Peptide Library Properties, 2014).<br>DNA Encodings: Number of oligonucleotide sequences possible within the defined encoding scheme that can encode the peptide.<br>Incl. Probability: Chance that the peptide will be part of the defined library."

tooltip_inclusion <- "The inclusion probability increases with the library size. Further the encoding scheme and thereby the number of possible encodings for a given peptide have a great influence. For 20/20 libraries that encode each amino acid with only one codon, all peptides have the same probability of inclusion in the library."
tooltip_neighborhood <- "The first degree neighborhood represents all peptides which differ in up to one amino acid positions from the peptide of interest. Allowed are only conservative exchanges as defined by a positive log-odds score in the BLOSOM80 matrix (Henikoff, S. and Henikoff, J. G. (1992). Amino acid substitution matrices from protein blocks. Proceedings of the National Academy of Sciences, 89(22), 10915–10919.). For instance the peptides YENNING and HENNING are in the degree 1 neighborhood of HENNING.<br><br>The second degree neighborhood represents all peptides which differ in up to two amino acid positions from the peptide of interest. Allowed are only conservative exchanges as defined by a positive log-odds score in the BLOSOM80 matrix (Henikoff, S. and Henikoff, J. G. (1992). Amino acid substitution matrices from protein blocks. Proceedings of the National Academy of Sciences, 89(22), 10915–10919.). For instance the peptides YENNING and QINNING are in the degree 2 neighborhood of HENNING."
tooltip_coverage <- "The coverage reflects the fraction of all theoretically possible peptide sequences that are encoded in your library."
tooltip_efficiency <- "The relative efficiency is a quality marker for your library that reflects the relation between the number of distinct valid peptides in the library and the total number of sequences at the plasmid/transformant level (N). For example, a library of 100 sequences that encodes only 50 distinct valid sequences has a relative efficiency of 0.5. As the chance that a random peptide added to the library is novel decreases with increasing library size, the relative efficiency of the library also decreases."
tooltip_diversity <- "Functional diversity is not a quantitative, but a qualitative measure of library diversity and mostly independent of the actual number of sequences in the library. It reflects if a library is biased to encoding some sequences more often than others (e.g. as consequence of the used encoding scheme, as shown here). The maximum value of \"1\" indicates a library in which every peptide has the same frequency. With increasingly skew distributions, this value drops towards a minimum of “0”. (Makowski, L. and Soares, A. (2003). Estimating the diversity of peptide populations from limited sequence data. Bioinformatics, 19(4), 483–489.; Rodi, D. J., Soares, A. S., and Makowski, L. (2002). Quantitative assessment of peptide sequence diversity in M13 combinatorial peptide phage display libraries. J. Mol. Biol., 322(5), 1039–1052.)"

text_welcome <- "PeLiCa is designed to facilitate the use of sophisticated mathematical methods to plan, analyze and evaluate your peptide libraries for viral (e.g. AAV), phage or other display systems.<br><br>PeLiCa allows you to:<br><br>-- Determine library diversity and coverage on a peptide level<br>-- Calculate the probabilities that top-performing peptides are included in the library<br>-- Examine the behaviour of individual peptides within the library<br><br>If you are using PeLiCa, please cite our <a href='http://www.ncbi.nlm.nih.gov/pubmed/26042419'>publication</a> and add the following link: http://www.pelica.org"
text_authors <- "<b>Timo Sieber</b>, University Medical Center Hamburg-Eppendorf, Department of Oncology and Hematology, 20246 Hamburg, Germany.<br><a href='http://www.erichare.net' target='_blank'><b>Eric Hare</b></a>, Department of Statistics, Iowa State University, Ames IA 50011-1210, United States.<br><b>Heike Hofmann</b>, Department of Statistics, Iowa State University, Ames IA 50011-1210, United States.<br><b>Martin Treppel</b>, University Medical Center Hamburg-Eppendorf, Department of Oncology and Hematology, 20246 Hamburg, Germany"
